x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/CompletionIT.java

```diff
@@ -0,0 +1,215 @@
+            for (List<Object> row : values) {
+                assertThat(row.get(1), notNullValue());
+            }
+        }
+    }
+
+    public void testCompletionWithCustomFieldName() {
+        var query = String.format(Locale.ROOT, """
+            FROM %s
+            | COMPLETION generated_text = title WITH { "inference_id": "%s" }
+            | KEEP id, generated_text
+            | LIMIT 3
+            """, TEST_INDEX, COMPLETION_MODEL_ID);
+
+        try (var resp = run(query)) {
+            assertColumnNames(resp.columns(), List.of("id", "generated_text"));
+            assertColumnTypes(resp.columns(), List.of("integer", "keyword"));
+
+            List<List<Object>> values = getValuesList(resp);
+            assertThat(values, hasSize(lessThanOrEqualTo(3)));
+        }
+    }
+
+    public void testCompletionConcatenatedPrompt() {
+        var query = String.format(Locale.ROOT, """
+            FROM %s
+            | COMPLETION summary = CONCAT(title, ": ", content) WITH { "inference_id": "%s" }
+            | KEEP id, summary
+            | LIMIT 5
+            """, TEST_INDEX, COMPLETION_MODEL_ID);
+
+        try (var resp = run(query)) {
+            assertColumnNames(resp.columns(), List.of("id", "summary"));
+
+            List<List<Object>> values = getValuesList(resp);
+            assertThat(values, hasSize(lessThanOrEqualTo(5)));
+
+            // Verify summary field is not null
+            for (List<Object> row : values) {
+                assertThat(row.get(1), notNullValue());
+            }
+        }
+    }
+
+    // ============================================
+    // Row Limit Tests
+    // ============================================
+
+    public void testCompletionRowLimitEnforcement() {
+        var query = String.format(Locale.ROOT, """
+            FROM %s
+            | COMPLETION title WITH { "inference_id": "%s" }
+            | KEEP id, completion
+            """, TEST_INDEX, COMPLETION_MODEL_ID);
+
+        try (var resp = run(query)) {
+            List<List<Object>> values = getValuesList(resp);
+            // Should be limited by the default row limit (100)
+            assertThat(values.size(), lessThanOrEqualTo(100));
+        }
+    }
+
+    public void testCompletionRowLimitSetting() throws Exception {
+        // Set a custom row limit
+        int customLimit = between(1, 10);
+        updateClusterSettings(Settings.builder().put(InferenceSettings.COMPLETION_ROW_LIMIT_SETTING.getKey(), customLimit));
+
+        try {
+            var query = String.format(Locale.ROOT, """
+                FROM %s
+                | COMPLETION title WITH { "inference_id": "%s" }
+                | KEEP id, completion
+                """, TEST_INDEX, COMPLETION_MODEL_ID);
+
+            try (var resp = run(query)) {
+                List<List<Object>> values = getValuesList(resp);
+                assertThat(values.size(), lessThanOrEqualTo(customLimit));
```

Comment:
let's check we get exactly `customLimit` docs back (take a look at the other comment that suggests changing `createAndPopulateTestIndex`.